### PR TITLE
Add support for cache reencryption

### PR DESCRIPTION
**Pull request by iambmelt**
Impetus:
- https://identitydivision.visualstudio.com/Engineering/_workitems/edit/1300836
- https://identitydivision.visualstudio.com/Engineering/_workitems/edit/1274352

Related Design PR:
- https://identitydivision.visualstudio.com/DevEx/_git/AuthLibrariesApiReview/pullrequest/2638

Change summary:
- Provides a high-level API inside of `common` to support an in-place reencryption of an instance of `ISharedPreferencesFileManager`

### CODE REVIEW

File: common/src/main/java/com/microsoft/identity/common/internal/cache/DefaultSharedPrefsFileManagerReencrypter.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+ * Default implementation of {@link ISharedPrefsFileManagerReencrypter}.
+ */
+public class DefaultSharedPrefsFileManagerReencrypter implements ISharedPrefsFileManagerReencrypter {
+
+    @Override
+    public void reencrypt(@NonNull final ISharedPreferencesFileManager fileManager,
+                          @NonNull final IStringEncrypter encrypter,
+                          @NonNull final IStringDecrypter decrypter,
+                          @NonNull final ReencryptionParams params) throws Exception {
+        final Map<String, String> cacheEntries = new HashMap<>(fileManager.getAll());
```
Comment (AdamBJohnsonx) [inline common/src/main/java/com/microsoft/identity/common/internal/cache/DefaultSharedPrefsFileManagerReencrypter.java:41]: I hate to ask, but what's guarding access while all this is going on?  I think it's all up to use behavior, make sure that while you are running the re-encryption, you don't insert any new keys?

Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/internal/cache/DefaultSharedPrefsFileManagerReencrypter.java:41]: You got it -- this is going to have to be guidanced-based. There's currently no way to programmatically enforce we have exclusive domain over the underlying `SharedPreferences` instance

Comment (AdamBJohnsonx) [reply common/src/main/java/com/microsoft/identity/common/internal/cache/DefaultSharedPrefsFileManagerReencrypter.java:41]: Do you want to be really explicit about this in the API documentation, then?

**Review by AdamBJohnsonx — APPROVED** [review]

File: common/src/main/java/com/microsoft/identity/common/internal/cache/DefaultSharedPrefsFileManagerReencrypter.java (hunk 1 above)
Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/internal/cache/DefaultSharedPrefsFileManagerReencrypter.java:41]: Roger wilco

Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/internal/cache/DefaultSharedPrefsFileManagerReencrypter.java:41]: 69c7eb6a46aefd9fd3b96d21ca9606ff15536d3e

**Pull request merged**